### PR TITLE
Add language support to the CRF API

### PR DESCRIPTION
**Pull request by psiyan**
- Change the API spec to account for the same.
   The new API has the following format:

   ```json
   {
     "language1": [
       {"sentence": "string1", "entities": ["tag1", "tag2"]},
       {"sentence": "string2", "entities": ["tag3", "tag4"]}
     ],
     "language2": [
       {"sentence": "string3", "entities": ["tag5", "tag6"]},
       {"sentence": "string4", "entities": ["tag7", "tag8"]}
     ]
   }
   ```

 - Change the delete query in CRF, it now accounts for languages and
   does a direct delete query instead of the existing fetch-scroll-bulk
   delete.
 - Segregated the above delete method from the existing delete method.
 - Renamed CRF only methods to specify the term `crf` in their name.
 - Improved documentation.
 - Changed old `%s` style string formatting in loggers to new .format
 - Minor changes in indentation.

### CODE REVIEW

File: datastore/datastore.py [hunk 1]
```diff
@@ -628,19 +629,18 @@ def get_crf_data_for_entity_name(self, entity_name, **kwargs):
             ner_logger.debug('Datastore, get_entity_training_data, results_dictionary %s' % str(entity_name))
         return results_dictionary
 
-    def update_entity_crf_data(self, entity_name, entity_list, language_script, sentence_list, **kwargs):
+    def update_entity_crf_data(self, entity_name, sentences: Dict[str, List], **kwargs):
```
Comment (srajdev) [inline datastore/datastore.py]: E999 SyntaxError: invalid syntax

File: datastore/elastic_search/populate.py [hunk 2]
```diff
@@ -280,83 +281,127 @@ def entity_data_update(connection, index_name, doc_type, entity_data, entity_nam
         logger.debug('%s: +++ Completed: add_data_elastic_search() +++' % log_prefix)
 
 
-def update_entity_crf_data_populate(
-    connection, index_name, doc_type, entity_list, entity_name, sentence_list, language_script, logger, **kwargs
-):
+def delete_entity_crf_data(connection: Elasticsearch, index_name: str,
```
Comment (srajdev) [inline datastore/elastic_search/populate.py]: E999 SyntaxError: invalid syntax

File: datastore/elastic_search/populate.py [hunk 3]
```diff
@@ -280,83 +281,127 @@ def entity_data_update(connection, index_name, doc_type, entity_data, entity_nam
...
+                           doc_type: str, entity_name: str,
+                           languages: List[str]):
+    """Delete CRF data for the given entity and languages.
+
```
Comment (srajdev) [inline datastore/elastic_search/populate.py]: W293 blank line contains whitespace

File: datastore/elastic_search/populate.py [hunk 4]
```diff
@@ -280,83 +281,127 @@ def entity_data_update(connection, index_name, doc_type, entity_data, entity_nam
...
+                           languages: List[str]):
+    """Delete CRF data for the given entity and languages.
+    
+    Args:
+        connection (Elasticsearch): Elasticsearch client object
+        index_name (str): name of the index
+        doc_type (str): type of the documents being indexed
+        entity_name (str):  ame of the entity for which the training data has to be deleted
+        languages (List[str]): list of language codes for which data needs to be deleted
+
```
Comment (srajdev) [inline datastore/elastic_search/populate.py]: W293 blank line contains whitespace

File: datastore/elastic_search/populate.py [hunk 5]
```diff
@@ -280,83 +281,127 @@ def entity_data_update(connection, index_name, doc_type, entity_data, entity_nam
...
+    logger.debug(f'[{log_prefix}] Started: delete_entity_crf_data()')
+    languages = list(sentences.keys())
+    delete_entity_crf_data(connection=connection, index_name=index_name, doc_type=doc_type,
+                           entity_name=entity_name, languages=languages)
+    logger.debug(f'[{log_prefix}] Completed: delete_entity_crf_data()')
+
+    logger.debug(f'[{log_prefix}] Started: add_training_data_elastic_search()')
+    add_crf_training_data_elastic_search(connection=connection,
+                                     index_name=index_name,
```
Comment (srajdev) [inline datastore/elastic_search/populate.py]: E128 continuation line under-indented for visual indent

File: datastore/elastic_search/populate.py [hunk 6]
```diff
@@ -280,83 +281,127 @@ def entity_data_update(connection, index_name, doc_type, entity_data, entity_nam
...
+                                     doc_type=doc_type,
```
Comment (srajdev) [inline datastore/elastic_search/populate.py]: E128 continuation line under-indented for visual indent

File: datastore/elastic_search/populate.py [hunk 7]
```diff
@@ -280,83 +281,127 @@ def entity_data_update(connection, index_name, doc_type, entity_data, entity_nam
...
                                      entity_name=entity_name,
-                                     entity_list=entity_list,
-                                     sentence_list=sentence_list,
-                                     language_script=language_script, logger=logger, **kwargs)
-    logger.debug('%s: +++ Completed: add_training_data_elastic_search() +++' % log_prefix)
+                                     sentences=sentences,
```
Comment (srajdev) [inline datastore/elastic_search/populate.py]: E128 continuation line under-indented for visual indent

File: datastore/elastic_search/populate.py [hunk 8]
```diff
@@ -280,83 +281,127 @@ def entity_data_update(connection, index_name, doc_type, entity_data, entity_nam
...
+                                     logger=logger, **kwargs)
```
Comment (srajdev) [inline datastore/elastic_search/populate.py]: E128 continuation line under-indented for visual indent

File: datastore/datastore.py [hunk 9]
```diff
@@ -628,19 +629,19 @@ def get_crf_data_for_entity_name(self, entity_name, **kwargs):
             ner_logger.debug('Datastore, get_entity_training_data, results_dictionary %s' % str(entity_name))
         return results_dictionary
 
-    def update_entity_crf_data(self, entity_name, entity_list, language_script, sentence_list, **kwargs):
+    def update_entity_crf_data(self, entity_name,
+                               sentences: Dict[str, List[Dict[str, Union[str, List[str]]]]], **kwargs):
```
Comment (srajdev) [inline datastore/datastore.py]: E999 SyntaxError: invalid syntax

File: datastore/datastore.py [hunk 10]
```diff
@@ -575,37 +576,40 @@ def transfer_entities_elastic_search(self, entity_list):
         es_object = elastic_search.transfer.ESTransfer(source=es_url, destination=destination)
         es_object.transfer_specific_entities(list_of_entities=entity_list)
 
-    def get_crf_data_for_entity_name(self, entity_name, **kwargs):
+    def get_crf_data_for_entity_name(self, entity_name: str, languages: List[str], **kwargs):
```
Comment (srajdev) [inline datastore/datastore.py]: E999 SyntaxError: invalid syntax

File: datastore/elastic_search/query.py [hunk 11]
```diff
@@ -533,17 +535,18 @@ def _parse_es_search_results(results_list):
     return variants_to_values_list
 
 
-def get_crf_data_for_entity_name(connection, index_name, doc_type, entity_name, **kwargs):
+def get_crf_data_for_entity_name(connection: Elasticsearch, index_name: str, doc_type: str, entity_name: str,
```
Comment (srajdev) [inline datastore/elastic_search/query.py]: E999 SyntaxError: invalid syntax

File: datastore/elastic_search/query.py [hunk 12]
```diff
@@ -564,15 +567,25 @@ def get_crf_data_for_entity_name(connection, index_name, doc_type, entity_name,
                 'Ajay'
             ]
                         ]
-            }
-
+            }
```
Comment (srajdev) [inline datastore/elastic_search/query.py]: W291 trailing whitespace

File: datastore/datastore.py [hunk 13]
```diff
@@ -628,19 +629,19 @@ def get_crf_data_for_entity_name(self, entity_name, **kwargs):
             ner_logger.debug('Datastore, get_entity_training_data, results_dictionary %s' % str(entity_name))
         return results_dictionary
 
-    def update_entity_crf_data(self, entity_name, entity_list, language_script, sentence_list, **kwargs):
+    def update_entity_crf_data(self, entity_name,
```
Comment (ashutoshsingh0223) [inline datastore/datastore.py]: Add type hint for `entity_name`

File: datastore/datastore.py [hunk 14]
```diff
@@ -1,6 +1,7 @@
 from __future__ import absolute_import
 
 import warnings
+from typing import Dict, List, Union
```
Comment (srajdev) [inline datastore/datastore.py]: F401 'typing.Union' imported but unused
F401 'typing.List' imported but unused
F401 'typing.Dict' imported but unused

File: datastore/elastic_search/populate.py [hunk 15]
```diff
@@ -3,11 +3,12 @@
 # std imports
 import os
 from collections import defaultdict
+# Local imports
+from typing import List, Dict, Union
```
Comment (srajdev) [inline datastore/elastic_search/populate.py]: F401 'typing.Union' imported but unused
F401 'typing.Dict' imported but unused
F401 'typing.List' imported but unused

File: datastore/elastic_search/populate.py [hunk 16]
```diff
@@ -3,11 +3,12 @@
...
 
 # 3rd party imports
-from elasticsearch import helpers
+from elasticsearch import helpers, Elasticsearch
```
Comment (srajdev) [inline datastore/elastic_search/populate.py]: F401 'elasticsearch.Elasticsearch' imported but unused

File: datastore/elastic_search/query.py [hunk 17]
```diff
@@ -6,10 +6,12 @@
 import json
 import re
 import warnings
+# Local imports
+from typing import List
```
Comment (srajdev) [inline datastore/elastic_search/query.py]: F401 'typing.List' imported but unused

File: datastore/elastic_search/query.py [hunk 18]
```diff
@@ -6,10 +6,12 @@
...
 
+from elasticsearch import Elasticsearch
```
Comment (srajdev) [inline datastore/elastic_search/query.py]: F401 'elasticsearch.Elasticsearch' imported but unused

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by psiyan** [comment]
retest this please

**Comment by psiyan** [comment]
retest this please

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by srajdev** [comment]
:+1: No lint errors found.

File: datastore/elastic_search/query.py [hunk 19]
```diff
@@ -9,12 +9,13 @@
 
 from six import string_types
 
-# Local imports
 from datastore import constants
-from external_api.constants import SENTENCE_LIST, ENTITY_LIST
+from external_api.constants import SENTENCE_LIST, ENTITY_LIST, LANGUAGE_SCRIPT, SENTENCE
```
Comment (srajdev) [inline datastore/elastic_search/query.py]: F401 'external_api.constants.SENTENCE_LIST' imported but unused
F401 'external_api.constants.LANGUAGE_SCRIPT' imported but unused

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by srajdev** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Pull request merged**